Define Tor Zone, add onion addressing and more

SyntheticBird45 opened the pull request:
### What

Part of [ongoing Tor support development](https://ccs.getmonero.org/proposals/syntheticbird_cuprate_arti_2_months.html) for Cuprate

Follows: https://github.com/Cuprate/cuprate/pull/446

### Changelog

In `cuprated`:

- Added `Z: NetworkZone` as a generic requirement for `address_book_config`. Now takes the optional node address in argument.
- Added `tor_net_seed_nodes` fn for obtaining tor seed nodes.
- Added `CrossNetworkInternalPeerId::Tor(_)` variant and `From<InternalPeerId<OnionAddr>>`

In `cuprate-wire`:

- Added `src/network_address/onion_addr.rs` implementing `OnionAddr` type used by `Tor` network zone.
- Implemented parsing, formatting, conversion and validation of `OnionAddr`.
- Implemented 2 validation tests and 2 parsing tests for `OnionAddr`.
- Documented and cleaned up `src/network_address/epee_builder.rs`.
- Changed `u8` `type` field of `TaggedNetworkAddress` to `AddressType` enum equivalent to monerod's `address_type`.
- Added additional `host` and `port` fields to `AllFieldedNetworkAddress` collection type.
- Added `NetworkAddress:Tor` variant and added conversion to related functions.

In `cuprate-address-book`:

- Added `our_own_addr: Z::Addr` field to AddressBookConfig. This adds a `Z: NetworkZone` requirement to `AddressBookConfig`.
- Adapted code to the new generic requirement.
- Implemented handling of new `AddressBookRequest::OwnAddress` for querying the node self specified address for the zone.

In `cuprate-p2p`:

- If `Z::BROADCAST_OUR_OWN_ADDR` = `true`, `handle_timed_sync_request` will insert the node's address to the peerlist being sent.

In `cuprate-p2p-core`:

- Removed `#[async_trait::async_trait]` attribute to `impl NetworkZone for *`.
- Added `AddressBookRequest::OwnAddress` and `AddressBookResponse::OwnAddress(Option<Z::Addr>)`.
- Defined new `Tor` `NetworkZone`

SyntheticBird45 commented on binaries/cuprated/src/config/p2p.rs:
```diff
@@ -335,3 +353,25 @@ pub fn clear_net_seed_nodes(network: Network) -> Vec<SocketAddr> {
+/// Seed nodes for `Tor`.
+pub fn tor_net_seed_nodes(network: Network) -> Vec<OnionAddr> {
+    let seeds = match network {
+        Network::Mainnet => [
+            "zbjkbsxc5munw3qusl7j2hpcmikhqocdf4pqhnhtpzw5nt5jrmofptid.onion:18083",
+            "qz43zul2x56jexzoqgkx2trzwcfnr6l3hbtfcfx54g4r3eahy3bssjyd.onion:18083",
+            "plowsof3t5hogddwabaeiyrno25efmzfxyro2vligremt7sxpsclfaid.onion:18083",
+            "plowsoffjexmxalw73tkjmf422gq6575fc7vicuu4javzn2ynnte6tyd.onion:18083",
+            "plowsofe6cleftfmk2raiw5h2x66atrik3nja4bfd3zrfa2hdlgworad.onion:18083",
+            "aclc4e2jhhtr44guufbnwk5bzwhaecinax4yip4wr4tjn27sjsfg6zqd.onion:18083",
+        ]
```
https://github.com/monero-project/monero/blob/master/src/p2p/net_node.inl#L888-L895

SyntheticBird45 commented on net/wire/src/network_address.rs:
```diff
@@ -106,7 +112,7 @@ impl TryFrom<NetworkAddress> for SocketAddr {
         match value {
             NetworkAddress::Clear(addr) => Ok(addr),
-            //_ => Err(NetworkAddressIncorrectZone)
+            NetworkAddress::Tor(_) => Err(NetworkAddressIncorrectZone),
```
just in case putting `_ =>` here causes a clippy error because `Tor` is the only variant remaining.

SyntheticBird45 commented on net/wire/src/network_address/epee_builder.rs:
```diff
@@ -1,21 +1,116 @@
+//! Address epee serialization
+//!
+//! Addresses needs to be serialized into a specific format before being sent to other peers.
+//! This module is handling this particular construction.
+//!
+
```
Did some cleanup here on the way. `cuprate-wire` is pretty important but didn't received much love.

SyntheticBird45 commented on net/wire/src/network_address/epee_builder.rs:
```diff
@@ -1,21 +1,116 @@
+    /// IPv6 address
+    addr: Option<[u8; 16]>,
+
+    /// Alternative network domain name (<domain>.onion or <domain>.i2p)
+    host: Option<String>,
+    /// Alternative network virtual port
+    port: Option<u16>,
```
Two new fields collected for alternative networks.

SyntheticBird45 commented on net/wire/src/network_address/epee_builder.rs:
```diff
@@ -1,21 +1,116 @@
+
+//---------------------------------------------------------------------------------------------------- Network Address Tag
+
+// Tag indicating the address type.
+// <https://github.com/monero-project/monero/blob/341771ac3e65f88a6e78ca805f8dde2ae6ba7924/contrib/epee/include/net/enums.h#L39>
+#[repr(u8)]
+enum NetworkAddressTag {
+    IPv4 = 1,
+    IPv6 = 2,
+    //I2p = 3,
+    Tor = 4,
+}
```
Defined an enum type instead of direct number for clarity.

Boog900 replied:
IMHO I do not think this is necessary. This is too Rusty.

Just keeping it `u8` is wayy simpler here + we don't need to expose other API from epee-encoding

SyntheticBird45 replied:
I would prefer it to be defined in a rusty way still, I didn't like the magic number + comments. @hinto-janai thoughts?

SyntheticBird45 replied:
As precised in meeting, replaced with already existing `AddressType`.

SyntheticBird45 commented on net/wire/src/network_address/onion_addr.rs:
```diff
@@ -0,0 +1,224 @@
+//! Onion address
+//!
+//! This module define v3 tor onion addresses
+//!
+//!
```
This is the brand new `OnionAddr` type that is simply port + 56 character domain name. It needed to implement `Copy` so I ended up using a `[u8; 56]` (a `[u8; 62]` for including .onion would have been useless)

SyntheticBird45 commented on net/wire/src/network_address/onion_addr.rs:
```diff
@@ -0,0 +1,224 @@
+    }
+}
+
+/// Display for [`OnionAddr`] only print the **onion address**. Not the port.
+impl Display for OnionAddr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```
`Display` will print the domain only: `<domain>.onion`

Boog900 replied:
ports do mean something for Tor though, right?

SyntheticBird45 replied:
Display is implemented for using `to_string()` in `epee_builder.rs` which really just need the onion address, not the port (as it is put as u16 in port field). I could define a `addr_string()` for OnionAddr and add port to `Display` ?

Boog900 replied:
> I could define a addr_string() for OnionAddr and add port to Display ?

Yes please

SyntheticBird45 commented on net/wire/src/network_address/onion_addr.rs:
```diff
@@ -0,0 +1,224 @@
+    }
+}
+
+/// [`OnionAddr`] parses an onion address **and a port**.
+impl FromStr for OnionAddr {
+    type Err = OnionAddrParsingError;
```
`FromStr` or `parse()` will attempt to parse domain + port: `<domain>.onion:<port>`

SyntheticBird45 commented on net/wire/src/network_address/onion_addr.rs:
```diff
@@ -0,0 +1,224 @@
+    #[test]
+    fn valid_onion_address() {
+        for addr in VALID_ONION_ADDRESSES {
+            assert!(
+                OnionAddr::check_addr(addr).is_ok(),
+                "Address {addr} has been reported as invalid."
+            );
+        }
+    }
+
+    #[test]
+    fn invalid_onion_address() {
+        for addr in INVALID_ONION_ADDRESSES {
+            assert!(
+                OnionAddr::check_addr(addr).is_err(),
+                "Address {addr} has been reported as valid."
+            );
+        }
+    }
+
+    #[test]
+    fn parse_valid_onion_address_w_port() {
+        for addr in VALID_ONION_ADDRESSES_WITH_PORT {
+            assert!(
+                addr.parse::<OnionAddr>().is_ok(),
+                "Address {addr} has been reported as invalid."
+            );
+        }
+    }
+
+    #[test]
+    fn parse_invalid_onion_address_w_port() {
+        for addr in INVALID_ONION_ADDRESSES_WITH_PORT {
+            assert!(
+                addr.parse::<OnionAddr>().is_err(),
+                "Address {addr} has been reported as valid."
+            );
+        }
+    }
+}
```
Added tests for validation.

Boog900 replied:
seems like a good place for a simple `proptest`

SyntheticBird45 commented on p2p/address-book/src/lib.rs:
```diff
@@ -33,6 +33,9 @@ pub struct AddressBookConfig {
     pub peer_store_directory: PathBuf,
     /// The amount of time between saving the address book to disk.
     pub peer_save_period: Duration,
+
+    /// Our own address to advertise to peers. (Only set if `Z::BROADCAST_OWN_ADDR` = `true`)
+    pub our_own_address: Option<Z::Addr>,
```
`AddressBookConfig` now contains our self-designated anonymous address. It is first fetched at config time in `cuprated` and can optionally be added at runtime by say `Arti`.

This makes `AddressBookConfig` now `AddressBookConfig<Z: NetworkZone>`.

SyntheticBird45 commented on p2p/p2p-core/src/client/request_handler.rs:
```diff
@@ -101,24 +103,58 @@ where
+            .call(CoreSyncDataRequest)
+            .await?;
+
+        // Attempt to fetch our own address if supported by this network zone.
+        let own_addr = if Z::BROADCAST_OWN_ADDR {
+            let AddressBookResponse::OwnAddress(own_addr) = self
+                .address_book_svc
+                .ready()
+                .await?
+                .call(AddressBookRequest::OwnAddress)
+                .await?
+            else {
+                panic!("Address book sent incorrect response!");
+            };
+
+            own_addr
+        } else {
+            None
+        };
+
+        let mut peer_list_req_size = MAX_PEERS_IN_PEER_LIST_MESSAGE;
+        if own_addr.is_some() {
+            peer_list_req_size -= 1;
+        }
```
similar to monerod we're propagating our anonymous address through the `local_peerlist_new` field of timed sync response.

SyntheticBird45 commented on p2p/p2p-core/src/client/request_handler.rs:
```diff
@@ -101,24 +103,58 @@ where
-            .await?;
+        if let Some(own_addr) = own_addr {
+            // Append our address to the final peer list
+            peers.push(ZoneSpecificPeerListEntryBase {
+                adr: own_addr,
+                id: self.our_basic_node_data.peer_id,
+                last_seen: 0,
+                pruning_seed: PruningSeed::NotPruned,
+                rpc_port: self.our_basic_node_data.rpc_port,
+                rpc_credits_per_hash: self.our_basic_node_data.rpc_credits_per_hash,
+            });
```
It's appended here.

Boog900 replied:
this should be inserted at a random position, or the vec should be reshuffled.

SyntheticBird45 commented on p2p/p2p-core/src/network_zones/tor.rs:
```diff
@@ -0,0 +1,52 @@
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct Tor;
+
+impl NetworkZone for Tor {
+    const NAME: &'static str = "Tor";
+
+    const CHECK_NODE_ID: bool = false;
+
+    const BROADCAST_OWN_ADDR: bool = true;
+
+    type Addr = OnionAddr;
+}
```
The new `Tor` netzone.

SyntheticBird45 commented on p2p/p2p-core/src/services.rs:
```diff
@@ -175,4 +178,10 @@ pub enum AddressBookResponse<Z: NetworkZone> {
 
     /// Response to [`AddressBookRequest::GetBans`].
     GetBans(Vec<BanState<Z::Addr>>),
+
+    /// Response to [`AddressBookRequest::OwnAddress`]
+    ///
+    /// This returns [`None`] if the address book do
+    /// not contain a self designated address.
+    OwnAddress(Option<Z::Addr>),
```
New request for getting the self designated address in `AddressBookConfig`

Boog900 commented on binaries/cuprated/src/config/p2p.rs:
```diff
@@ -1,16 +1,18 @@
     time::Duration,
 };
 
+use cuprate_wire::OnionAddr;
 use serde::{Deserialize, Serialize};
```
:(

SyntheticBird45 replied:
oops.

Boog900 commented on net/wire/src/network_address/onion_addr.rs:
```diff
@@ -0,0 +1,224 @@
+            return Err(OnionAddrParsingError::InvalidTld(String::from(tld)));
+        }
+
+        // The domain part must only contain base32 characters.
+        if !domain
+            .as_bytes()
+            .iter()
+            .copied()
+            .all(|c| c.is_ascii_lowercase() || (b'2'..=b'7').contains(&c))
+        {
+            return Err(OnionAddrParsingError::NonBase32Char);
+        }
```
IMO this check is probably unneeded, people can still create invalid addresses even with this check and the only reason this would be invalid is because someone is actively being malicious.Even with this check if someone wanted to be malicious they can create invalid addresses.

SyntheticBird45 replied:
https://github.com/monero-project/monero/blob/master/src/net/tor_address.cpp#L63-L67

It is needed as monerod also reject addresses with invalid characters. If we don't reject them Cuprate will end up relaying invalid addresses.

Moreover, I just find it to be common sense to make sure we don't include invalid data and later assumes we're only dealing with API that doesn't care about it ("let's arti/tor resolver sort the bad apples").

It's only returning an error as of now but as you said this is most likely a malicious intent, this could later be considered punishable if repeated.

Boog900 commented on net/wire/src/network_address/onion_addr.rs:
```diff
@@ -0,0 +1,224 @@
+        // Port
+        let port: u16 = port
+            .starts_with(':')
+            .then_some(port[1..].parse().ok())
```
this will panic if the string is empty, i.e has nothing after the `.onion`.

Boog900 replied:
I don't think you ran the same test:

```
byte index 1 is out of bounds of ``
thread 'network_address::onion_addr::tests::parse_invalid_onion_address_w_port' panicked at net/wire/src/network_address/onion_addr.rs:111:28:
byte index 1 is out of bounds of ``
```

You have fallen into a pretty common trap, `then_some` is a function that takes in a value `T` and outputs it if the `bool` is `true`. The value `T` is computed before the call and passed. What you want is something that takes in a closure/function and runs that closure/function to produce the value `T` if the bool is `Some`.

```suggestion
            .then(|| port[1..].parse().ok())
```

Boog900 commented on net/wire/src/network_address/onion_addr.rs:
```diff
@@ -0,0 +1,224 @@
+            .ok_or(OnionAddrParsingError::InvalidPort)?;
+
+        // Address
+        Self::check_addr(addr)?;
+        let domain = addr.as_bytes()[..56]
+            .try_into()
+            .unwrap_or_else(|e| panic!("We just validated address: {addr} : {e}"));
```
`check_addr` could return the domain

Boog900 commented on p2p/p2p-core/src/client/request_handler.rs:
```diff
@@ -1,3 +1,4 @@
+use cuprate_pruning::PruningSeed;
 use futures::TryFutureExt;
```
:(

SyntheticBird45 commented on binaries/cuprated/src/config/p2p.rs:
```diff
@@ -228,20 +230,36 @@ config_struct! {
     }
 }
 
+/// Helper trait for fetching the config field holding the address of this particular zone.
+pub trait FetchOwnAddr<Z: NetworkZone> {
+    fn zone_own_addr(&self) -> Option<Z::Addr>;
+}
+
+impl FetchOwnAddr<ClearNet> for SharedNetConfig {
+    fn zone_own_addr(&self) -> Option<<ClearNet as NetworkZone>::Addr> {
+        // [`ClearNet::BROADCAST_OWN_ADDR`] = `false`
+        None
+    }
+}
```
This is to avoid `address_book_config_tor`, `address_book_config_clearnet` duplication now that `AddressBookConfig` takes `Z`, the trait bounds forces definition while letting flexibility in the future on how we fetch this address (config now, ControlPort maybe later on for Tor, even some different mechanism for I2P).

Boog900 commented on Cargo.toml:
```diff
@@ -110,6 +110,7 @@ arc-swap              = { version = "1", default-features = false }
 arrayvec              = { version = "0.7", default-features = false }
 async-trait           = { version = "0.1", default-features = false }
 bitflags              = { version = "2", default-features = false }
+base32                = { version = "0.5.1", default-features = false }
```
do we need this dep anymore?

Boog900 commented on Cargo.toml:
```diff
@@ -139,6 +140,7 @@ rayon                 = { version = "1", default-features = false }
 serde_bytes           = { version = "0.11", default-features = false }
 serde_json            = { version = "1", default-features = false }
 serde                 = { version = "1", default-features = false }
+sha3                  = { version = "0.10.8", default-features = false }
```
and this one

Boog900 commented on net/wire/src/network_address/epee_builder.rs:
```diff
@@ -1,21 +1,65 @@
+//---------------------------------------------------------------------------------------------------- Network address construction
+
+#[derive(Default)]
+/// There are no ordering guarantees in epee format and as such all potential fields can be collected during deserialization.
+/// The [`AllFieldsNetworkAddress`] is containing, as its name suggest, all optional field describing an address , as if it
+/// could be of any type.
+struct AllFieldsNetworkAddress {
```
Why has this been moved? 

Although here it isn't too bad, doing reshuffles like this complicates the diff so should be avoided.

Boog900 replied:
yeah different style, I normally go the other way round. I don't actually care which way too much but for existing code I wouldn't swap stuff around.

Boog900 commented on net/wire/src/network_address/onion_addr.rs:
```diff
@@ -0,0 +1,239 @@
+    ///
+    /// Returns a `String` containing the onion domain name and ".onion" TLD only, in form of `zbjkbs...ofptid.onion`.
+    pub fn addr_string(&self) -> String {
+        let mut str = self
+            .domain
+            .utf8_chunks()
+            .next()
+            .expect("Onion addresses are always containing UTF-8 characters.")
+            .valid()
+            .to_string();
```
can we not just use https://doc.rust-lang.org/std/str/fn.from_utf8.html here?

Boog900 commented on net/wire/src/network_address/onion_addr.rs:
```diff
@@ -0,0 +1,239 @@
+/// Display for [`OnionAddr`]. It prints the onion address and port, in the form of `<domain>.onion:<port>`
+impl Display for OnionAddr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let str = self
+            .domain
+            .utf8_chunks()
+            .next()
+            .expect("Onion addresses are always containing UTF-8 characters.");
```
and here

Boog900 reviewed (commented):
just some nits

SyntheticBird45 committed:
Define Tor Zone, add onion addressing, extend AddressBook, adapt `handle_timed_sync_request`, changes in cuprated + some cleanup

In `cuprated`:

- Added `Z: NetworkZone` as a generic requirement for `address_book_config`. Now takes the optional node address in argument.
- Added `tor_net_seed_nodes` fn for obtaining tor seed nodes.
- Added `CrossNetworkInternalPeerId::Tor(_)` variant and `From<InternalPeerId<OnionAddr>>`

In `cuprate-wire`:

- Added `src/network_address/onion_addr.rs` implementing `OnionAddr` type used by `Tor` network zone.
- Implemented parsing, formatting, conversion and validation of `OnionAddr`.
- Implemented 2 validation tests and 2 parsing tests for `OnionAddr`.
- Documented and cleaned up `src/network_address/epee_builder.rs`.
- Changed `u8` `type` field of `TaggedNetworkAddress` to `AddressType` enum equivalent to monerod's `address_type`.
- Added additional `host` and `port` fields to `AllFieldedNetworkAddress` collection type.
- Added `NetworkAddress:Tor` variant and added conversion to related functions.

In `cuprate-address-book`:

- Added `our_own_addr: Z::Addr` field to AddressBookConfig. This adds a `Z: NetworkZone` requirement to `AddressBookConfig`.
- Adapted code to the new generic requirement.
- Implemented handling of new `AddressBookRequest::OwnAddress` for querying the node self specified address for the zone.

In `cuprate-p2p`:

- If `Z::BROADCAST_OUR_OWN_ADDR` = `true`, `handle_timed_sync_request` will insert the node's address to the peerlist being sent.

In `cuprate-p2p-core`:

- Removed `#[async_trait::async_trait]` attribute to `impl NetworkZone for *`.
- Added `AddressBookRequest::OwnAddress` and `AddressBookResponse::OwnAddress(Option<Z::Addr>)`.
- Defined new `Tor` `NetworkZone`

SyntheticBird45 committed:
fmt

SyntheticBird45 committed:
Merge branch 'main' of github.com:Cuprate/cuprate into tor-zone

SyntheticBird45 committed:
fix typo and fmt

SyntheticBird45 committed:
final edits?

SyntheticBird45 committed:
fix test

SyntheticBird45 committed:
forgor
